pgp/keysource.go

```diff
@@ -0,0 +1,202 @@
+package pgp
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"github.com/howeyc/gopass"
+	gpgagent "go.mozilla.org/gopgagent"
+	"golang.org/x/crypto/openpgp"
+	"golang.org/x/crypto/openpgp/armor"
+	"io/ioutil"
+	"os"
+	"os/user"
+	"path"
+	"strings"
+	"time"
+)
+
+const DateFormat = "2006-01-02T15:04:05Z"
+
+type GPGMasterKey struct {
+	Fingerprint  string
+	EncryptedKey string
+	CreationDate time.Time
+}
+
+func (key *GPGMasterKey) Encrypt(dataKey string) error {
+	ring, err := key.pubRing()
+	if err != nil {
+		return err
+	}
+	fingerprints := key.fingerprintMap(ring)
+	entity, ok := fingerprints[key.Fingerprint]
+	if !ok {
+		return fmt.Errorf("Key with fingerprint %s is not available in keyring.", key.Fingerprint)
+	}
+	encbuf := new(bytes.Buffer)
+	armorbuf, err := armor.Encode(encbuf, "PGP MESSAGE", nil)
+	if err != nil {
+		return err
+	}
+	plaintextbuf, err := openpgp.Encrypt(armorbuf, []*openpgp.Entity{&entity}, nil, nil, nil)
+	if err != nil {
+		return err
+	}
+	_, err = plaintextbuf.Write([]byte(dataKey))
+	if err != nil {
+		return err
+	}
+	err = plaintextbuf.Close()
+	if err != nil {
+		return err
+	}
+	err = armorbuf.Close()
+	if err != nil {
+		return err
+	}
+	bytes, err := ioutil.ReadAll(encbuf)
+	if err != nil {
+		return err
+	}
+	key.EncryptedKey = string(bytes)
+	return nil
+}
+
+func (key *GPGMasterKey) EncryptIfNeeded(dataKey string) error {
+	if key.EncryptedKey == "" {
+		return key.Encrypt(dataKey)
+	}
+	return nil
+}
+
+func (key *GPGMasterKey) Decrypt() (string, error) {
+	ring, err := key.secRing()
+	if err != nil {
+		return "", fmt.Errorf("Could not load secring: %s", err)
+	}
+	block, err := armor.Decode(strings.NewReader(key.EncryptedKey))
+	if err != nil {
+		return "", fmt.Errorf("Armor decoding failed: %s", err)
+	}
+	md, err := openpgp.ReadMessage(block.Body, ring, key.passphrasePrompt, nil)
+	if err != nil {
+		return "", fmt.Errorf("Reading PGP message failed: %s", err)
+	}
+	if b, err := ioutil.ReadAll(md.UnverifiedBody); err == nil {
+		return string(b), nil
+	}
+	return "", fmt.Errorf("The key could not be decrypted with any of the GPG entries")
+}
+
+func (key *GPGMasterKey) NeedsRotation() bool {
+	return time.Since(key.CreationDate).Hours() > 24*30*6
```

Comment:
Another way to write this would be `time.Now().Add(-time.Hour*24*30*6).After(key.CreationDate)`
but that's not very readable...